woocommerce/woocommerce-ios
# [TTPReq] Integrate Education steps after Terms of Service are accepted

Closes: #14327


## Description


Integrate merchant education into TTP connection flow and present it after Terms of Service are accepted. Two main flows should be supported:

- TTP onboarding from "Payments -> Set up Tap to Pay"
- TTP onboarding during the Tap to Pay payment

More information: https://github.com/woocommerce/woocommerce-ios/issues/14327

## Solution

1. Use [localMobileReaderDidAcceptTermsOfService](https://github.com/stripe/stripe-terminal-ios/blob/master/PublicHeaders/SCPLocalMobileReaderDelegate.h#L107C9-L107C49) delegate event from Stripe to determine when TOS is accepted
2. Create  `builtInCardReaderAcceptToSEvents` publisher in `StripeCardReaderService` to emit an event when TOS is accepted.
3. Create `BuiltInCardReaderMerchantEducationPresenter` to present `TapToPayEducationViewViewHostingController` on top of already presented modals.
4. Update `BuiltInCardReaderConnectionController` to observe `builtInCardReaderOnboardingEvents` and present merchant education
5. Update `BuiltInCardReaderConnectionController` states to include `educationInProgress` in associated values. It is used to determine whether the connection process should complete or be set to `waitToComplete` if connection has finished but the education is still ongoing.

There are a few key decisions made here that could be approached in different ways:
- How `StripeCardReaderService` communicates that TOS is accepted
- How we use presenters/alert providers to present TapToPayEducationView
- How we interrupt the connection flow while merchant education is presented

## Steps to reproduce


Prerequisites between each test case:
- Enable `tapToPayEducation` in `DefaultFeatureFlagService`
- Remove your device to have a fresh TTP setup https://businessconnect.apple.com/taptopay/removeall
- Use eligible iPhone and store for Tap to Pay

### Test Case 1: Set up Tap to Pay

1. Open Menu -> Payments
2. Select Set Up Tap to Pay
3. Proceed through the flow
4. Accept Terms of Service when it's presented
5. Confirm Tap to Pay education appears (the education content is still hardcoded)
7. Complete the flow (Skip/Next -> Done)
8. Confirm the connection flow finishes
9. Try out payment
10. Confirm payment succeeds and education doesn't appear anymore

### Test Case 2: Payment

1. Open Menu -> Collect Payment
2. Add Custom Amount
3. Proceed with Tap to Pay
4. Accept Terms of Service when it's presented
5. Confirm Tap to Pay education appears (the education content is still hardcoded)
6. Complete the flow (Skip/Next -> Done)
7. Confirm the payment flow proceeds once the education is done

### Refunds

I will handle refund testing and any necessary changes in another task.

## Testing information


I only did manual testing. The one place that contains the main business logic for the flow is `BuiltInCardReaderConnectionController` which currently doesn't have any tests. Do you think it makes sense to start covering it with tests?

## Videos


### Set up Tap to Pay flow

https://github.com/user-attachments/assets/70e30962-5aab-48c5-a0a4-bed72fbfd329


### Payment Flow

https://github.com/user-attachments/assets/1543063f-5b3a-4e57-9057-9e0e7b11f2ff


---

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:
